File: napier/src/commonMain/kotlin/com/github/aakira/napier/Napier.kt
package com.github.aakira.napier

import kotlin.native.concurrent.ThreadLocal

@ThreadLocal
object Napier {

    enum class Level {
        VERBOSE,
        DEBUG,
        INFO,
        WARNING,
        ERROR,
        ASSERT,
    }

    private val baseArray = mutableListOf<Antilog>()

    fun base(antilog: Antilog) {
        baseArray.add(antilog)
    }

    fun isEnable(priority: Level, tag: String?) = baseArray.any { it.isEnable(priority, tag) }

    @PublishedApi
    internal fun rawLog(priority: Level, tag: String?, throwable: Throwable?, message: String?) {
        baseArray.forEach { it.rawLog(priority, tag, throwable, message) }
    }

    fun v(message: String, throwable: Throwable? = null, tag: String? = null) {
        log(Level.VERBOSE, tag, throwable, message)
    }

    fun v(message: () -> String, throwable: Throwable? = null, tag: String? = null) {
        log(Level.VERBOSE, tag, throwable, message)
    }

    fun i(message: String, throwable: Throwable? = null, tag: String? = null) {
        log(Level.INFO, tag, throwable, message)
    }

    fun i(message: () -> String, throwable: Throwable? = null, tag: String? = null) {
        log(Level.INFO, tag, throwable, message)
    }

    fun d(message: String, throwable: Throwable? = null, tag: String? = null) {
        log(Level.DEBUG, tag, throwable, message)
    }

    fun d(message: () -> String, throwable: Throwable? = null, tag: String? = null) {
        log(Level.DEBUG, tag, throwable, message)
    }

    fun w(message: String, throwable: Throwable? = null, tag: String? = null) {
        log(Level.WARNING, tag, throwable, message)
    }

    fun w(message: () -> String, throwable: Throwable? = null, tag: String? = null) {
        log(Level.WARNING, tag, throwable, message)
    }

    fun e(message: String, throwable: Throwable? = null, tag: String? = null) {
        log(Level.ERROR, tag, throwable, message)
    }

    fun e(message: () -> String, throwable: Throwable? = null, tag: String? = null) {
        log(Level.ERROR, tag, throwable, message)
    }

    fun wtf(message: String, throwable: Throwable? = null, tag: String? = null) {
        log(Level.ASSERT, tag, throwable, message)
    }

    fun wtf(message: () -> String, throwable: Throwable? = null, tag: String? = null) {
        log(Level.ASSERT, tag, throwable, message)
    }

    fun log(priority: Level, tag: String? = null, throwable: Throwable? = null, message: String) {
        if (isEnable(priority, tag)) {
            rawLog(priority, tag, throwable, message)
        }
    }

    fun log(priority: Level, tag: String? = null, throwable: Throwable? = null, message: () -> String) {
        if (isEnable(priority, tag)) {
            rawLog(priority, tag, throwable, message())
        }
    }

    /**
     * Remove antilog from the base array.
     */
    fun takeLogarithm(antilog: Antilog) {
        baseArray.remove(antilog)
    }

    /**
     * Clear all antilogs from the base array.
     */
    fun takeLogarithm() {
        baseArray.clear()
    }
}

File: ios/Napier/Napier.swift
//
//  Napier.swift
//  Napier
//
//  Created by Phil on 23.10.2021.
//  Copyright © 2021 AAkira. All rights reserved.
//

import mpp_sample

extension Napier {
    static func v(tag: String? = nil, _ items: Any..., separator: String = " ", file: String = #file, function: String = #function) {
        log(logLevel: .verbose, tag: tag, items, separator: separator, file: file, function: function)
    }
    
    static func d(tag: String? = nil, _ items: Any..., separator: String = " ", file: String = #file, function: String = #function) {
        log(logLevel: .debug, tag: tag, items, separator: separator, file: file, function: function)
    }
    
    static func i(tag: String? = nil, _ items: Any..., separator: String = " ", file: String = #file, function: String = #function) {
        log(logLevel: .info, tag: tag, items, separator: separator, file: file, function: function)
    }
    
    static func w(tag: String? = nil, _ items: Any..., separator: String = " ", file: String = #file, function: String = #function) {
        log(logLevel: .warning, tag: tag, items, separator: separator, file: file, function: function)
    }
    
    static func e(tag: String? = nil, _ items: Any..., separator: String = " ", file: String = #file, function: String = #function) {
        log(logLevel: .error, tag: tag, items, separator: separator, file: file, function: function)
    }
    
    static func a(tag: String? = nil, _ items: Any..., separator: String = " ", file: String = #file, function: String = #function) {
        log(logLevel: .assert, tag: tag, items, separator: separator, file: file, function: function)
    }
    
    static private func log(logLevel: LogLevel, tag: String?, _ items: [Any], separator: String, file: String, function: String) {
        let message = items.map { "\($0)" }.joined(separator: separator)
        shared.log(
            priority: logLevel,
            tag: tag ?? {
                let fileName = URL(fileURLWithPath: file).lastPathComponent
                let functionName: String
                if let firstBraceIndex = function.firstIndex(of: "(") {
                    functionName = String(function[..<firstBraceIndex])
                } else {
                    functionName = function
                }
                return "\(fileName):\(functionName)"
            }(),
            throwable: nil,
            message: message
        )
    }
}
